### DISCUSSION TITLE
1.21

### DISCUSSION THREAD

**Discussion by ruskariimi** [sebseb7/autotrade-fabric#9] [General]
Any plans on updating for version 1.21? If not, what are some alternatives?

**Comment by TimonH-98** [comment]
I cant find a replacement for this Mod. I hope it gets updated soon. I need more quartzblocks xD

**Comment by basil222** [comment]
I followed this easy tutorial to update the mod form 1.20.4 to 1.21.3, and it works great(probably bc nothing really changed to villager trading) : 
https://www.youtube.com/watch?v=GWB8loTTfw0&t=11s

**Reply by TimonH-98** [reply]
This doesnt work for me sadly. Minecraft starts up and it seems to work on some servers, but on others i get a crash the instant i activate the autotrade. In the logs it says something about malilib and hotkeys. 
Super Sadness

**Comment by Kreys31** [comment]
Without any knowledge of Java, driven by the need for this mod for my specific use case (after the version where I only modified the manifest.json suddenly started crashing after using it without issues for some time),

I decided to take the easiest yet most complex path: forcibly updating the mod. With the help of my programming logic knowledge, lots of artificial intelligence, Fabric documentation, and thousands of iterations, I managed to update the mod by fixing all outdated Fabric API calls and several other things.

I haven’t tested it much—only one day of stable use. This is pretty unorthodox; it’s basically a Frankenstein. I’m not sure if the functions for selecting trades with item frames work, nor if containers with tinted glass function properly.

But the basic functionality—trading when near a villager, choosing an item to sell, setting the maximum price, and configuring input and output containers, along with the delay to close those containers—does work. At least when only “autotrade” and “malilib” are installed.

Which is enough for my specific use case. However, I’m not sure how far this mod adaptation works, whether it works on other 1.21 versions by modifying the manifest.json, or what incompatibilities may arise with other mods. You can try it yourselves if anyone out there is as desperate as I am to use this Frankenstein.

Details: For some reason, to sell at a max price of 1, I have to set it to 10, and to sell at a max price of 10, I have to set it to 15. Try it out yourselves.
Probably if I looked carefully and modified a few lines of code, I could fix this, but I’m fine with it as is.
Also, I’m a Spanish speaker, so you’ll find comments in Spanish and probably a poor translation of this post.

I'm sharing the project's source code with the "functional" .jar file precompiled inside, so you can verify that it's not malicious, compile it yourself if you wish, or even try to fix bugs and make improvements.
It’s compiled for Fabric 1.21.8, and I’m using it alongside Malilib 0.25.5 and FabricAPI 0.136.1.

You can compile it yourself by running "./gradlew remapJar" and "./gradlew build" in the project folder (Linux)
I believe all the necessary files are there. Although I might be mistaken.

I hope I can help someone as desperate as I was, and that the original developer won't be upset about this :p

[autotrade-fabric-master.zip](https://github.com/user-attachments/files/24034459/autotrade-fabric-master.zip)

**Reply by TimonH-98** [reply]
Thank you so much! I was able to update it with some help from ChatGPT to MC Version 1.21.10!
For the mod to work you need:
MC 1.21.10
Fabric Version 0.17.3
Malilib 0.26.6

The Mod Menü Mod broke the AFK Traiding Mod. So i just removed it. You can open the Mod settings in mc with the Hotkey Right shift + T. 
I never tried all funktions of the mod. I use villagers for quartz trading and that works! The traiding is way slower compared to older versions. But hey it works!


[autotrade-fabric 1.21.10.zip](https://github.com/user-attachments/files/24060151/autotrade-fabric.1.21.10.zip)

**Reply by Kreys31** [reply]
Update: 0.0.11.1

I noticed that the inconsistency with the SellLimit (and presumably the same applied to BuyLimit) was due to the system always using the base trade price as reference to decide if a trade was too expensive, rather than comparing the actual current price.

For example, in my case, I had to set SellLimit to at least 4 to be able to sell iron, because that's the base price for the iron trade. This meant that no matter how much the villager increased the price, it would always trade iron, even at 20 or more iron for a single emerald.

Apparently this was caused by incorrect use of the Fabric API which was retrieving the base value. I fixed it by finding the correct method call that gets the special price (the actual variable price based on reputation with the villager).

Then, for some reason, the trading interface with villagers who didn't meet the trade requirements wouldn't close, which prevented trading with new villagers until moving far enough away from the villager with the open interface.

I made it so the system detects when a trade is actually performed after interacting with a villager, and if no trade occurs, it closes the interface.

I hope that covers everything for now and no further issues arise, and that you can use this as a foundation for updating to 1.21.10 if this was causing problems for you.

[autotrade-fabric-master.zip](https://github.com/user-attachments/files/24085295/autotrade-fabric-master.zip)

Same compilation instructions and version specifications.

**Comment by Hyperlinkhyper1** [comment]
will there be a 1.21.11 version too, or can smh help me make it?

**Comment by majiangla** [comment]
我将它升级到了1.21.1
[autotrade-fabric-1.21.1-0.0.11.zip](https://github.com/user-attachments/files/26162589/autotrade-fabric-1.21.1-0.0.11.zip)
